pkg/auth/dependency/principal/password/normalize.go

```diff
@@ -0,0 +1,135 @@
+package password
+
+import (
+	"strings"
+
+	"golang.org/x/net/idna"
+	"golang.org/x/text/cases"
+	"golang.org/x/text/unicode/norm"
+
+	"github.com/skygeario/skygear-server/pkg/core/auth/metadata"
+	"github.com/skygeario/skygear-server/pkg/core/config"
+)
+
+type LoginIDNormalizar interface {
+	Normalize(loginID string) (string, error)
+	GenerateUniqueKey(normalizeLoginID string) (string, error)
+}
+
+type LoginIDNormalizarFactory interface {
+	NewNormalizar(loginIDKey string) LoginIDNormalizar
+}
+
+func NewLoginIDNormalizarFactory(
+	loginIDsKeys []config.LoginIDKeyConfiguration,
+	loginIDTypes *config.LoginIDTypesConfiguration,
+) LoginIDNormalizarFactory {
+	return &factoryImpl{
+		loginIDsKeys: loginIDsKeys,
+		loginIDTypes: loginIDTypes,
+	}
+}
+
+type factoryImpl struct {
+	loginIDsKeys []config.LoginIDKeyConfiguration
+	loginIDTypes *config.LoginIDTypesConfiguration
+}
+
+func (f *factoryImpl) NewNormalizar(loginIDKey string) LoginIDNormalizar {
+	for _, c := range f.loginIDsKeys {
+		if c.Key == loginIDKey {
+			return f.newNormalizar(c.Type)
+		}
+	}
+
+	panic("password: invalid login id key: " + loginIDKey)
+}
+
+func (f *factoryImpl) newNormalizar(loginIDKeyType config.LoginIDKeyType) LoginIDNormalizar {
+	metadataKey, _ := loginIDKeyType.MetadataKey()
+	switch metadataKey {
+	case metadata.Email:
+		return &LoginIDEmailNormalizar{
+			config: f.loginIDTypes.Email,
+		}
+	case metadata.Username:
+		return &LoginIDUsernameNormalizar{
+			config: f.loginIDTypes.Username,
+		}
+	}
+
+	return &LoginIDNullNormalizar{}
+}
+
+type LoginIDEmailNormalizar struct {
+	config *config.LoginIDTypeEmailConfiguration
+}
+
+func (n *LoginIDEmailNormalizar) Normalize(loginID string) (string, error) {
+	parts := strings.Split(loginID, "@")
+	local, domain := parts[0], parts[1]
+
+	// convert the domain part
+	c := cases.Fold()
```

Comment:
Docs of `cases.Fold` shows me that https://godoc.org/golang.org/x/text/secure/precis exists. Should we use that instead?